# Project 3: Explore Dataset of Used Cars from the German eBay Website

The dataset was originally scraped and uploaded to Kaggle. We've made a few modifications from the original dataset that was uploaded to Kaggle:

We sampled 50,000 data points from the full dataset, to ensure your code runs quickly in our hosted environment
We dirtied the dataset a bit to more closely resemble what you would expect from a scraped dataset (the version uploaded to Kaggle was cleaned to be easier to work with)

In [1]:
import pandas as pd
import numpy as np
autos = pd.read_csv('autos.csv', encoding='Latin-1')

In [2]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [3]:
# convert the column names from camelcase to snakecase 
# and reword some of the column names based on the data dictionary 

autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [4]:
col = autos.columns.str.replace('yearOfRegistration', 'registration_year')
col = col.str.replace('monthOfRegistration', 'registration_month')
col = col.str.replace('notRepairedDamage', 'unrepaired_damage')
col = col.str.replace('dateCreated', 'ad_created')
col

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'registration_year', 'gearbox', 'powerPS', 'model',
       'odometer', 'registration_month', 'fuelType', 'brand',
       'unrepaired_damage', 'ad_created', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [5]:
autos.columns = col
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'registration_year', 'gearbox', 'powerPS', 'model',
       'odometer', 'registration_month', 'fuelType', 'brand',
       'unrepaired_damage', 'ad_created', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [6]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  registration_year  \
0  privat   Angebot  $5,000  control         bus               2004   
1  privat   Angebot  $8,500  control   limousine               1997   
2  privat   Angebot  $8,990     test   limousine               2009   
3  privat   Angebot  $4,350  control  kleinwagen               2007   
4  privat   Angebot  $1,350     test       kombi               2003   

     gearbox  powerPS   model   odometer  registration_month fuelType  \
0    manuell      158  andere  150,000km                   3      lpg   
1  automatik      286     7er  150,000km                   6   benzin   
2    manuell      102    golf   70,000km                   7   benzin   
3  automatik       71  fortwo   70,000km                   6   benzin   
4    manuell        0   focus  150,000km                   7   benzin   

        brand unrepaired_damage           ad_created  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

I extract the column names and revised four of the names.

## 3. some basic data exploration to determine what other cleaning tasks need to be done

In [7]:
autos.describe()

registration_year       powerPS  registration_month  nrOfPictures  \
count       50000.000000  50000.000000        50000.000000       50000.0   
mean         2005.073280    116.355920            5.723360           0.0   
std           105.712813    209.216627            3.711984           0.0   
min          1000.000000      0.000000            0.000000           0.0   
25%          1999.000000     70.000000            3.000000           0.0   
50%          2003.000000    105.000000            6.000000           0.0   
75%          2008.000000    150.000000            9.000000           0.0   
max          9999.000000  17700.000000           12.000000           0.0   

         postalCode  
count  50000.000000  
mean   50813.627300  
std    25779.747957  
min     1067.000000  
25%    30451.000000  
50%    49577.000000  
75%    71540.000000  
max    99998.000000

abnormal facts: registration_month and powerPS min is 0. nrOfPictures column has all 0s.

In [8]:
autos['registration_month'].value_counts()

0     5075
3     5071
6     4368
5     4107
4     4102
7     3949
10    3651
12    3447
9     3389
11    3360
1     3282
8     3191
2     3008
Name: registration_month, dtype: int64

In [9]:
autos['powerPS'].value_counts()

0        5500
75       3171
60       2195
150      2046
140      1884
101      1756
90       1746
116      1646
170      1492
105      1410
125       964
136       955
102       868
163       847
54        759
143       733
131       713
122       710
110       694
109       620
50        604
80        560
177       542
58        506
120       501
115       481
69        475
45        397
95        382
68        380
         ... 
999         1
455         1
442         1
1082        1
678         1
454         1
187         1
262         1
441         1
585         1
460         1
5867        1
9011        1
268         1
236         1
1367        1
24          1
12          1
1771        1
1003        1
587         1
696         1
952         1
1016        1
682         1
650         1
490         1
362         1
153         1
16312       1
Name: powerPS, Length: 448, dtype: int64

In [10]:
autos['nrOfPictures'].value_counts()

0    50000
Name: nrOfPictures, dtype: int64

In [11]:
autos = autos.drop('nrOfPictures', axis=1)

In [12]:
autos.describe(include='all')

dateCrawled         name  seller offerType  price abtest  \
count                 50000        50000   50000     50000  50000  50000   
unique                48213        38754       2         2   2357      2   
top     2016-03-09 11:54:38  Ford_Fiesta  privat   Angebot     $0   test   
freq                      3           78   49999     49999   1421  25756   
mean                    NaN          NaN     NaN       NaN    NaN    NaN   
std                     NaN          NaN     NaN       NaN    NaN    NaN   
min                     NaN          NaN     NaN       NaN    NaN    NaN   
25%                     NaN          NaN     NaN       NaN    NaN    NaN   
50%                     NaN          NaN     NaN       NaN    NaN    NaN   
75%                     NaN          NaN     NaN       NaN    NaN    NaN   
max                     NaN          NaN     NaN       NaN    NaN    NaN   

       vehicleType  registration_year  gearbox       powerPS  model  \
count        44905       50000.000000    47320  50000.000000  47242   
unique           8                NaN        2           NaN    245   
top      limousine                NaN  manuell           NaN   golf   
freq         12859                NaN    36993           NaN   4024   
mean           NaN        2005.073280      NaN    116.355920    NaN   
std            NaN         105.712813      NaN    209.216627    NaN   
min            NaN        1000.000000      NaN      0.000000    NaN   
25%            NaN        1999.000000      NaN     70.000000    NaN   
50%            NaN        2003.000000      NaN    105.000000    NaN   
75%            NaN        2008.000000      NaN    150.000000    NaN   
max            NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuelType       brand unrepaired_damage  \
count       50000        50000.000000    45518       50000             40171   
unique         13                 NaN        7          40                 2   
top     150,000km                 NaN   benzin  volkswagen              nein   
freq        32424                 NaN    30107       10687             35232   
mean          NaN            5.723360      NaN         NaN               NaN   
std           NaN            3.711984      NaN         NaN               NaN   
min           NaN            0.000000      NaN         NaN               NaN   
25%           NaN            3.000000      NaN         NaN               NaN   
50%           NaN            6.000000      NaN         NaN               NaN   
75%           NaN            9.000000      NaN         NaN               NaN   
max           NaN           12.000000      NaN         NaN               NaN   

                 ad_created    postalCode             lastSeen  
count                 50000  50000.000000                50000  
unique                   76           NaN                39481  
top     2016-04-03 00:00:00           NaN  2016-04-07 06:17:27  
freq                   1946           NaN                    8  
mean                    NaN  50813.627300                  NaN  
std                     NaN  25779.747957                  NaN  
min                     NaN   1067.000000                  NaN  
25%                     NaN  30451.000000                  NaN  
50%                     NaN  49577.000000                  NaN  
75%                     NaN  71540.000000                  NaN  
max                     NaN  99998.000000                  NaN

e.g. price and odometer data are stored as text. Let's convert them to numeric data type

In [13]:

autos['price'] = (autos['price']
                  .str.replace('$','')
                  .str.replace(',', '')
                  .astype(int)
                  )
autos['price'].head()

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int64

In [14]:
autos['odometer'] = (autos['odometer']
                  .str.replace('km','')
                  .str.replace(',', '')
                  .astype(int)
                  )

autos.rename({"odometer": "odometer_km"}, axis=1, inplace=True)

In [15]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType  price   abtest vehicleType  registration_year    gearbox  \
0  privat   Angebot   5000  control         bus               2004    manuell   
1  privat   Angebot   8500  control   limousine               1997  automatik   
2  privat   Angebot   8990     test   limousine               2009    manuell   
3  privat   Angebot   4350  control  kleinwagen               2007  automatik   
4  privat   Angebot   1350     test       kombi               2003    manuell   

   powerPS   model  odometer_km  registration_month fuelType       brand  \
0      158  andere       150000                   3      lpg     peugeot   
1      286     7er       150000                   6   benzin         bmw   
2      102    golf        70000                   7   benzin  volkswagen   
3       71  fortwo        70000                   6   benzin       smart   
4        0   focus       150000                   7   benzin        ford   

  unrepaired_damage           ad_created  postalCode             lastSeen  
0              nein  2016-03-26 00:00:00       79588  2016-04-06 06:45:54  
1              nein  2016-04-04 00:00:00       71034  2016-04-06 14:45:08  
2              nein  2016-03-26 00:00:00       35394  2016-04-06 20:15:37  
3              nein  2016-03-12 00:00:00       33729  2016-03-15 03:16:28  
4              nein  2016-04-01 00:00:00       39218  2016-04-01 14:38:50

## 4. Next, we'll use:
- Series.unique().shape to see how many unique values
- Series.describe() to view min/max/median/mean etc
- Series.value_counts(), with some variations:
chained to .head() if there are lots of values.
Because Series.value_counts() returns a series, we can use Series.sort_index() with ascending= True or False to view the highest and lowest values with their counts (can also chain to head() here).
- When removing outliers, we can do df[(df["col"] > x ) & (df["col"] < y )], but it's more readable to use df[df["col"].between(x,y)]

In [16]:
autos['price'].unique()

array([ 5000,  8500,  8990, ...,   385, 22200, 16995])

In [17]:
autos['odometer_km'].unique()

array([150000,  70000,  50000,  80000,  10000,  30000, 125000,  90000,
        20000,  60000,   5000, 100000,  40000])

In [18]:
autos['price'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [19]:
autos['odometer_km'].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [20]:
autos['price'].value_counts().sort_index(ascending=False).head(10)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
Name: price, dtype: int64

In [21]:
autos['odometer_km'].value_counts().sort_index(ascending=False).head(10)

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
40000       819
30000       789
Name: odometer_km, dtype: int64

In [22]:
autos = autos[(autos['price']>10000) & (autos['price']<100000)]
autos['price'].describe()

count     8020.000000
mean     18893.743267
std      10010.435399
min      10009.000000
25%      12500.000000
50%      15880.000000
75%      21692.500000
max      99900.000000
Name: price, dtype: float64

In [23]:
autos = autos[autos['odometer_km'].between(5000, 100000)]
autos['odometer_km'].describe()

count      4747.000000
mean      57762.797556
std       28311.973221
min        5000.000000
25%       30000.000000
50%       60000.000000
75%       80000.000000
max      100000.000000
Name: odometer_km, dtype: float64

## 5. 

In [24]:
autos.rename({'dateCrawled': 'date_crawled'}, axis=1, inplace=True)
autos.rename({'lastSeen':'last_seen'}, axis=1, inplace=True)

In [25]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
15  2016-04-01 12:06:20  2016-04-01 00:00:00  2016-04-02 21:10:48
22  2016-03-28 20:50:54  2016-03-28 00:00:00  2016-04-01 06:45:30
24  2016-04-03 11:57:02  2016-04-03 00:00:00  2016-04-07 13:16:50
37  2016-03-23 16:47:46  2016-03-23 00:00:00  2016-04-05 20:17:53
42  2016-03-22 21:37:03  2016-03-22 00:00:00  2016-04-06 18:44:39

In [26]:
(autos['date_crawled']
 .str[:10].value_counts(normalize=True, dropna=False)
 .sort_index()
)
#str[:10]: 10 represent the number of characters in date                                                                     

2016-03-05    0.029071
2016-03-06    0.013061
2016-03-07    0.033916
2016-03-08    0.025700
2016-03-09    0.030335
2016-03-10    0.031599
2016-03-11    0.032231
2016-03-12    0.032442
2016-03-13    0.020223
2016-03-14    0.035391
2016-03-15    0.034127
2016-03-16    0.029703
2016-03-17    0.027596
2016-03-18    0.014325
2016-03-19    0.039604
2016-03-20    0.040236
2016-03-21    0.035812
2016-03-22    0.034337
2016-03-23    0.030756
2016-03-24    0.030546
2016-03-25    0.030335
2016-03-26    0.030124
2016-03-27    0.030967
2016-03-28    0.034127
2016-03-29    0.032863
2016-03-30    0.032442
2016-03-31    0.034337
2016-04-01    0.038972
2016-04-02    0.037708
2016-04-03    0.042975
2016-04-04    0.038551
2016-04-05    0.012008
2016-04-06    0.002107
2016-04-07    0.001475
Name: date_crawled, dtype: float64

Looks like the site was crawled daily over roughly a one month period in March and April 2016. The distribution of listings crawled on each day is roughly uniform.

In [27]:
(autos['ad_created']
 .str[:10].value_counts(normalize=True, dropna=False)
 .sort_index()
)

2015-06-11    0.000211
2016-01-03    0.000211
2016-01-22    0.000211
2016-02-05    0.000211
2016-02-07    0.000211
2016-02-17    0.000211
2016-02-19    0.000211
2016-02-23    0.000421
2016-02-25    0.000211
2016-02-28    0.000421
2016-02-29    0.000211
2016-03-02    0.000211
2016-03-03    0.000421
2016-03-04    0.001685
2016-03-05    0.026543
2016-03-06    0.013693
2016-03-07    0.032231
2016-03-08    0.026122
2016-03-09    0.031178
2016-03-10    0.030756
2016-03-11    0.032020
2016-03-12    0.033495
2016-03-13    0.021698
2016-03-14    0.033916
2016-03-15    0.032863
2016-03-16    0.030335
2016-03-17    0.027807
2016-03-18    0.014325
2016-03-19    0.038551
2016-03-20    0.040447
2016-03-21    0.036655
2016-03-22    0.033284
2016-03-23    0.030124
2016-03-24    0.030546
2016-03-25    0.030546
2016-03-26    0.030335
2016-03-27    0.031388
2016-03-28    0.033916
2016-03-29    0.032442
2016-03-30    0.033074
2016-03-31    0.034337
2016-04-01    0.038340
2016-04-02    0.037919
2016-04-03 

The starting date is from 2015 which is older than the other two columns.

In [28]:
(autos['last_seen']
 .str[:10].value_counts(normalize=True, dropna=False)
 .sort_index()
)

2016-03-05    0.000632
2016-03-06    0.001685
2016-03-07    0.001475
2016-03-08    0.001685
2016-03-09    0.004003
2016-03-10    0.005688
2016-03-11    0.006320
2016-03-12    0.008637
2016-03-13    0.004845
2016-03-14    0.009901
2016-03-15    0.010533
2016-03-16    0.009058
2016-03-17    0.018959
2016-03-18    0.005477
2016-03-19    0.009269
2016-03-20    0.013693
2016-03-21    0.014957
2016-03-22    0.017485
2016-03-23    0.013482
2016-03-24    0.012640
2016-03-25    0.011165
2016-03-26    0.010954
2016-03-27    0.008637
2016-03-28    0.010322
2016-03-29    0.013904
2016-03-30    0.022119
2016-03-31    0.017274
2016-04-01    0.017695
2016-04-02    0.020013
2016-04-03    0.024015
2016-04-04    0.019591
2016-04-05    0.168738
2016-04-06    0.297240
2016-04-07    0.187908
Name: last_seen, dtype: float64

The crawler recorded the date it last saw any listing, which allows us to determine on what day a listing was removed, presumably because the car was sold.

The last three days contain a disproportionate amount of 'last seen' values. Given that these are 6-10x the values from the previous days, it's unlikely that there was a massive spike in sales, and more likely that these values are to do with the crawling period ending and don't indicate car sales.

In [29]:
autos['registration_year'].describe()

count    4747.000000
mean     2013.539920
std       165.849195
min      1927.000000
25%      2009.000000
50%      2011.000000
75%      2013.000000
max      9999.000000
Name: registration_year, dtype: float64

The year that the car was first registered will likely indicate the age of the car. Looking at this column, we note some odd values. The minimum value is 1000, long before cars were invented and the maximum is 9999, many years into the future.

## 6. Dealing with Incorrect Registration Year Data

Because a car can't be first registered before the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate. Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s.

In [30]:
# ~ means not; .shape[0] gives the number of rows
(~autos["registration_year"].between(1900,2016)).sum() / autos.shape[0]

0.016010111649462817

In [31]:
autos = autos[autos['registration_year'].between(1900, 2016)]
autos["registration_year"].value_counts(normalize=True).head(10)

2012    0.156926
2011    0.156069
2013    0.123314
2014    0.116891
2010    0.097410
2009    0.071291
2015    0.070649
2008    0.053094
2007    0.030400
2006    0.020766
Name: registration_year, dtype: float64

## 7. Exploring Price by Brand

In [32]:
autos['brand'].unique()

array(['sonstige_autos', 'mini', 'bmw', 'seat', 'volkswagen', 'audi',
       'mercedes_benz', 'jaguar', 'ford', 'opel', 'chevrolet', 'porsche',
       'hyundai', 'honda', 'nissan', 'mitsubishi', 'renault', 'peugeot',
       'kia', 'skoda', 'toyota', 'volvo', 'dacia', 'fiat', 'alfa_romeo',
       'citroen', 'mazda', 'lancia', 'suzuki', 'jeep', 'chrysler',
       'smart', 'land_rover', 'subaru', 'saab'], dtype=object)

In [33]:
autos['brand'].value_counts(normalize=True)

volkswagen        0.216014
audi              0.149433
bmw               0.134661
mercedes_benz     0.132520
opel              0.045386
ford              0.044744
mini              0.035538
porsche           0.024620
skoda             0.023764
sonstige_autos    0.023121
nissan            0.017555
seat              0.016271
renault           0.014344
hyundai           0.012417
mazda             0.010918
toyota            0.010062
kia               0.009848
peugeot           0.009848
chevrolet         0.007921
citroen           0.007279
fiat              0.007279
volvo             0.005994
land_rover        0.005780
jeep              0.004710
honda             0.004710
suzuki            0.004282
alfa_romeo        0.004282
mitsubishi        0.004068
dacia             0.003425
jaguar            0.003211
subaru            0.001713
chrysler          0.001713
smart             0.001285
lancia            0.000642
saab              0.000642
Name: brand, dtype: float64

In [34]:
brand_counts = autos['brand'].value_counts(normalize=True)
big_brand = brand_counts[brand_counts>0.05].index

In [35]:
big_brand

Index(['volkswagen', 'audi', 'bmw', 'mercedes_benz'], dtype='object')

In [36]:
mean_price = {}
for b in big_brand:
    rows = autos[autos['brand']==b]
    avg_price = rows['price'].mean()
    mean_price[b]=int(avg_price)
print(mean_price)

{'mercedes_benz': 23895, 'bmw': 22220, 'volkswagen': 18364, 'audi': 23799}


## 8. Exploring Mileage

In [41]:
bmp_series = pd.Series(mean_price)
price_mean = pd.DataFrame(bmp_series, columns=["mean_price"])
price_mean

mean_price
audi                23799
bmw                 22220
mercedes_benz       23895
volkswagen          18364

In [42]:
mean_mileage = {}
for b in big_brand:
    rows = autos[autos['brand']==b]
    avg_mileage = rows['odometer_km'].mean()
    mean_mileage[b]=int(avg_mileage)

bmp_series = pd.Series(mean_mileage)
pd.DataFrame(bmp_series, columns=["mean_mileage"])

mean_mileage
audi                  58051
bmw                   66096
mercedes_benz         63432
volkswagen            57319

In [43]:
brand_info = pd.DataFrame(bmp_series, columns=["mean_mileage"])
brand_info['mean_price'] = price_mean
brand_info

mean_mileage  mean_price
audi                  58051       23799
bmw                   66096       22220
mercedes_benz         63432       23895
volkswagen            57319       18364